### PR TITLE
Initial Codec implementations that will be used to serialized thrift structures to zk

### PR DESCRIPTION
**Pull request by gopalrajpurohit**

### CODE REVIEW

File: cluster_management/src/test/java/com/pinterest/rocksplicator/codecs/ThriftCodecTest.java [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+public class ThriftCodecTest extends CodecTestBase {
+
+  @Test
+  public void testBinaryCodec() throws Exception {
+    ThriftCodec<LeaderEventsHistory>
+        codec =
+        ThriftCodec.createBinaryCodec(LeaderEventsHistory.class);
+    byte[] binaryData = codec.encode(history);
+    LeaderEventsHistory decodedHistory = codec.decode(binaryData);
+    assertEquals(history, decodedHistory);
```
Comment (rajathprasad) [inline cluster_management/src/test/java/com/pinterest/rocksplicator/codecs/ThriftCodecTest.java:18]: I'm missing this in all the tests, but where is this variable `history` declared?

File: cluster_management/src/test/java/com/pinterest/rocksplicator/codecs/WrappedDataThriftCodecTest.java [hunk 2]
```diff
@@ -0,0 +1,97 @@
+package com.pinterest.rocksplicator.codecs;
```
Comment (rajathprasad) [inline cluster_management/src/test/java/com/pinterest/rocksplicator/codecs/WrappedDataThriftCodecTest.java:1]: Also, are these tests run as part of the Travis CI?

**Review by rajathprasad — APPROVED** [review]

File: cluster_management/src/test/java/com/pinterest/rocksplicator/codecs/WrappedDataThriftCodecTest.java [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+        wrappedCodecCU =
+        new WrappedDataThriftCodec(LeaderEventsHistory.class, SerializationProtocol.BINARY,
+            CompressionAlgorithm.UNCOMPRESSED);
+    WrappedDataThriftCodec<LeaderEventsHistory>
+        wrappedCodecCG =
+        new WrappedDataThriftCodec(LeaderEventsHistory.class, SerializationProtocol.BINARY,
+            CompressionAlgorithm.GZIP);
+    WrappedDataThriftCodec<LeaderEventsHistory>
+        wrappedCodecCS =
+        new WrappedDataThriftCodec(LeaderEventsHistory.class, SerializationProtocol.BINARY,
```
[the remark this answers is not in the document]
Comment (gopalrajpurohit) [reply cluster_management/src/test/java/com/pinterest/rocksplicator/codecs/WrappedDataThriftCodecTest.java:39]: Thanks. Yes, they should be using COMPACT protocol.

File: cluster_management/src/main/java/com/pinterest/rocksplicator/codecs/WrappedDataThriftCodec.java [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+  final Class<S> thriftClazz;
+  final SerializationProtocol serializationProtocol;
+  final CompressionAlgorithm compressionAlgorithm;
+  final Map<SerializationProtocol, Map<CompressionAlgorithm, Codec<S, byte[]>>> allCodecsPairs;
+  private Codec<WrappedData, byte[]> wrappedDataBinaryCodec;
+
+  public WrappedDataThriftCodec(
+      final Class<S> thriftClazz,
+      final SerializationProtocol serializationProtocol,
+      final CompressionAlgorithm compressionAlgorithm) {
```
[the remark this answers is not in the document]
Comment (gopalrajpurohit) [reply cluster_management/src/main/java/com/pinterest/rocksplicator/codecs/WrappedDataThriftCodec.java:24]: encode is a method from interface Codec. Codec doesn't have concept of Thrift. It is generic interface. WrappedDataThriftCodec is thrift specific codec implementation. Hence it needs to be configured to use one or the other.

File: cluster_management/src/test/java/com/pinterest/rocksplicator/codecs/ThriftCodecTest.java (hunk 1 above)
Comment (gopalrajpurohit) [reply cluster_management/src/test/java/com/pinterest/rocksplicator/codecs/ThriftCodecTest.java:18]: In the CodecTestBase class which is Base class for all codec tests.

File: cluster_management/src/test/java/com/pinterest/rocksplicator/codecs/WrappedDataThriftCodecTest.java (hunk 2 above)
Comment (gopalrajpurohit) [reply cluster_management/src/test/java/com/pinterest/rocksplicator/codecs/WrappedDataThriftCodecTest.java:1]: I don't think so. the travis yaml as defined is 

- cd ../cluster_management; mvn clean package -DskipTests # if you want to test:  mvn "-DfailIfNoTests=false" test
which will skip the tests. I tested them locally using mvn clean package in cluster_management directory. I also used local zk and performed writes and reads from zk using these codecs..

**Pull request merged**